From b0db0fe36175187775a956f63d2a8707df99d4e8 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:10:30 -0400
Subject: [powerpc] le: Avoid creatng R_PPC64_TOCSAVE relocations for modules

Message-id: <1410545655-205645-402-git-send-email-dzickus@redhat.com>
Patchwork-id: 94588
O-Subject: [RHEL7 PATCH 401/626] powerpc/le: Avoid creatng R_PPC64_TOCSAVE relocations for modules.
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit d3d35d957a9d0733dc51f14b5abc0bff5d3c5f3a
Author: Tony Breeds <tony@bakeyournoodle.com>
Date:   Wed Mar 12 15:12:01 2014 +1100

    powerpc/le: Avoid creatng R_PPC64_TOCSAVE relocations for modules.

    When building modules with a native le toolchain the linker will
    generate R_PPC64_TOCSAVE relocations when it's safe to omit saving r2 on
    a plt call.  This isn't helpful in the conext of a kernel module and the
    kernel will fail to load those modules with an error like:
     nf_conntrack: Unknown ADD relocation: 109

    This patch tells the linker to avoid createing R_PPC64_TOCSAVE
    relocations allowing modules to load.

    Signed-off-by: Tony Breeds <tony@bakeyournoodle.com>
    Signed-off-by: Anton Blanchard <anton@samba.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/Makefile b/arch/powerpc/Makefile
index e719e29..c0b031b 100644
--- a/arch/powerpc/Makefile
+++ b/arch/powerpc/Makefile
@@ -74,6 +74,7 @@ override CROSS32AS += -mlittle-endian
 LDEMULATION := lppc
 GNUTARGET := powerpcle
 MULTIPLEWORD := -mno-multiple
+KBUILD_CFLAGS_MODULE += $(call cc-option,-mno-save-toc-indirect)
 else
 ifeq ($(call cc-option-yn,-mbig-endian),y)
 override CC += -mbig-endian
-- 
1.7.1